### PR TITLE
support non-coroutine asgi apps in TestClient

### PR DESCRIPTION
**Pull request by adriangb**
This adds support for testing ASGI apps like `starlette.endpoints.HTTPEndpoint` via TestClient

This is a piece of #1453

### CODE REVIEW

File: starlette/testclient.py [hunk 1]
```diff
@@ -81,26 +79,22 @@ def _get_reason_phrase(status_code: int) -> str:
...
     async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
-        instance = self.app(scope)
-        await instance(receive, send)
+        try:
+            asgi3app = typing.cast(ASGI3App, self.app)
+            aw = asgi3app(scope, receive, send)
+        except TypeError:
+            asgi2app = typing.cast(ASGI2App, self.app)
+            aw = asgi2app(scope)(receive, send)
+        await aw
```
Comment (lovelydinosaur) [inline starlette/testclient.py:97]: Do we need to support ASGI 2.0 anymore?
Perhaps we could look at dropping that to keep things more obvious?

(I don't see any great resource in the ASGI docs on the 2 -> 3 changes?)

Comment (Kludex) [reply starlette/testclient.py:97]: here: https://asgi.readthedocs.io/en/latest/specs/main.html?highlight=2.0#legacy-applications

Comment (adriangb) [reply starlette/testclient.py:97]: Based on that it seems reasonable for TestClient to drop support for ASGI 2.0, I'm happy to drop it here or in another prequel/sequel PR if there is consensus. As is this is a backwards compatible change.

Comment (lovelydinosaur) [reply starlette/testclient.py:97]: I think dropping ASGI 2 first would be pretty reasonable, yup.

**Comment by adriangb** [comment]
I opened #1473 to remove ASGI2 support.

I also think that #1453 is an interesting experiment, but is likely not worth moving forward with.

Once #1473 gets merged and eliminates the issue false detection of non-coroutine ASGI 3 apps as ASGI 2.0 apps (whaat this PR was addressing) there is still a tiny issue, but it's on the AnyIO side and would be fixed by https://github.com/agronholm/anyio/pull/416.

Given that no one has complained about this thus far, I'm going to close this PR and if someone _does_ want to use TestClient with `HTTPEndpoint` directly, they'll just have to wait until https://github.com/agronholm/anyio/pull/416 gets merged.

**Pull request closed without merging**